x-pack/plugins/security_solution/public/common/components/markdown_editor/plugins/insight/index.tsx

```diff
@@ -0,0 +1,167 @@
+import { i18n } from '@kbn/i18n';
+import { useAppToasts } from '../../../../hooks/use_app_toasts';
+import { useInsightQuery } from './use_insight_query';
+import { useInsightDataProviders } from './use_insight_data_providers';
+import { BasicAlertDataContext } from '../../../event_details/investigation_guide_view';
+import { InvestigateInTimelineButton } from '../../../event_details/table/investigate_in_timeline_button';
+
+interface InsightComponentProps {
+  label?: string;
+  description?: string;
+  providers?: string;
+}
+
+export const parser: Plugin = function () {
+  const Parser = this.Parser;
+  const tokenizers = Parser.prototype.inlineTokenizers;
+  const methods = Parser.prototype.inlineMethods;
+  const insightPrefix = '!{insight';
+
+  const tokenizeInsight: RemarkTokenizer = function (eat, value, silent) {
+    if (value.startsWith(insightPrefix) === false) {
+      return false;
+    }
+
+    const nextChar = value[insightPrefix.length];
+    if (nextChar !== '{' && nextChar !== '}') return false;
+    if (silent) {
+      return true;
+    }
+
+    // is there a configuration?
+    const hasConfiguration = nextChar === '{';
+
+    let configuration: InsightComponentProps = {};
+    if (hasConfiguration) {
+      let configurationString = '';
+      let openObjects = 0;
+
+      for (let i = insightPrefix.length; i < value.length; i++) {
+        const char = value[i];
+        if (char === '{') {
+          openObjects++;
+          configurationString += char;
+        } else if (char === '}') {
+          openObjects--;
+          if (openObjects === -1) {
+            break;
+          }
+          configurationString += char;
+        } else {
+          configurationString += char;
+        }
+      }
+
+      try {
+        configuration = JSON.parse(configurationString);
+        return eat(value)({
+          type: 'insight',
+          ...configuration,
+          providers: JSON.stringify(configuration.providers),
+        });
+      } catch (err) {
+        const now = eat.now();
+        this.file.fail(
+          i18n.translate('xpack.securitySolution.markdownEditor.plugins.insightConfigError', {
+            values: { err },
+            defaultMessage: 'Unable to parse insight JSON configuration: {err}',
+          }),
+          {
+            line: now.line,
+            column: now.column + insightPrefix.length,
+          }
+        );
+      }
+    }
+    return false;
+  };
+  tokenizeInsight.locator = (value: string, fromIndex: number) => {
+    return value.indexOf(insightPrefix, fromIndex);
+  };
+  tokenizers.insight = tokenizeInsight;
+  methods.splice(methods.indexOf('text'), 0, 'insight');
+};
+
+// receives the configuration from the parser and renders
+const InsightComponent = ({ label, description, providers }: InsightComponentProps) => {
+  const { addError } = useAppToasts();
+  let parsedProviders = [];
+  try {
+    if (providers !== undefined) {
```

Comment:
providers can never be null?

Reply:
if they are, we just pass an empty array for data providers, and the button will render but show 0